### PR TITLE
Convert mod_articles_archive to new structure

### PR DESCRIPTION
**Pull request by carlitorweb**
### Summary of Changes
This PR converts mod_articles_archive to the new structure

### Testing Instructions
- Use Joomla 4.3
- Create an instance of the module Articles - Archived. Note how it works.
- Apply patch
- Delete administrator/cache/autoload_psr4.php
- Refresh the page, make sure the module still displays the same as before.

### Additional comments
A small change to the layout was made. Now will show the amount of articles the month have archived, for example:

**Before patch:**
December, 2022

**After patch:**
December, 2022 (1)

If is not allowed, let me kown and I revert the change

### CODE REVIEW

**Comment by carlitorweb** [comment]
@laoneo the name of the method `getArticles()` do not describe really the purpose of the method. Can be changed for this module to something like `getMonths()` for example?

File: modules/mod_articles_archive/tmpl/default.php [hunk 1]
```diff
@@ -19,7 +19,7 @@
     <?php foreach ($list as $item) : ?>
     <li>
         <a href="<?php echo $item->link; ?>">
-            <?php echo $item->text; ?>
+            <?php echo $item->text; ?> (<?php echo $item->amount; ?>)
```
Comment (ReLater) [inline modules/mod_articles_archive/tmpl/default.php]: Should the amount get an accessible description/label or be a badge with "Article Count:" or similar?
@brianteeman

Comment (carlitorweb) [reply modules/mod_articles_archive/tmpl/default.php]: @ReLater I though in add something like "Article Count:" when doing the changes, but as this _change_ is not related with the purpose of this PR, even when is a small change, I added only the counter just in case. But yes, should have some description and not only a number who really not have a context

Comment (ReLater) [reply modules/mod_articles_archive/tmpl/default.php]: I don't know if there are any a11y rules for changes like this. I just wanted to mention it because I didn't understand what "amount" means for the first moment.

**Comment by laoneo** [comment]
Sure, change it. More accurate names are always better...

File: modules/mod_articles_archive/tmpl/default.php (hunk 1 above)
Comment (chmst) [reply modules/mod_articles_archive/tmpl/default.php]: Thanks for the hint, @ReLater. It is surely an UX issue and maybe an a11y issue. But not in scope of this pr.

Comment (chmst) [reply modules/mod_articles_archive/tmpl/default.php]: Maybe it is better to make an own correct PR for that when this one is merged and disply no amount here, @carlitorweb ?

Comment (carlitorweb) [reply modules/mod_articles_archive/tmpl/default.php]: @chmst Sure. Even change "amount" key if is needed, look how @ReLater did not understand what "amount" means for the first moment. So for sure need another word.

Comment (ReLater) [reply modules/mod_articles_archive/tmpl/default.php]: I personally don't care and don't know if there are rules for a11y like for CS indention errors that would block merging. Just for reference (NOT MORE) how a badge could look like:
https://github.com/joomla/joomla-cms/blob/4.3.0-alpha2/components/com_content/tmpl/categories/default_items.php#L29-L32

Comment (carlitorweb) [reply modules/mod_articles_archive/tmpl/default.php]: @ReLater sorry i said it in the good way. I wanted meant, if you who are someone used to the Joomla code, did not catch the meaning of amount, we need change it then. Sorry, sometimes my English is not good enough for express my opinion. All is okay, and I agree with you. For sure will change it.

Comment (brianteeman) [reply modules/mod_articles_archive/tmpl/default.php]: sorry for the delay in replying - I thought you were saying that you did not understand the code $amount - we use that in the code over 1000 times

Comment (ReLater) [reply modules/mod_articles_archive/tmpl/default.php]: To clarify: I didn't understand amount in code ;-) but also didn't understand the number `(3)` on page display ;-) But please ignore my comments and set it to status "resolved"! Too much noise at the moment! It's not worth it as long as we can override the output.

Comment (carlitorweb) [reply modules/mod_articles_archive/tmpl/default.php]: I will change it. In fact I always wanted change even $list and $item, who really not describe nothing in this context. Better will be $months and $month. Then use $month->numarticles.

**Comment by carlitorweb** [comment]
@obuisard is normal that commits are included here?

**Comment by obuisard** [comment]
> @obuisard is normal that commits are included here?

I have no idea how it got here. Guided Tours PR has nothing to do with this PR.

File: modules/mod_articles_archive/tmpl/default.php [hunk 2]
```diff
@@ -10,16 +10,17 @@
...
-if (!$list) {
+/** @var  \stdClass[]  $months  Archived articles by months */
+if (!$months) {
     return;
 }
 
 ?>
 <ul class="mod-articlesarchive archive-module mod-list">
-    <?php foreach ($list as $item) : ?>
+    <?php foreach ($months as $month) : ?>
```
Comment (laoneo) [inline modules/mod_articles_archive/tmpl/default.php]: This changes here must be reverted, otherwise it will break template overrides.

**Review by laoneo — APPROVED** [review]

File: modules/mod_articles_archive/tmpl/default.php [hunk 3]
```diff
@@ -19,7 +20,7 @@
     <?php foreach ($list as $item) : ?>
     <li>
         <a href="<?php echo $item->link; ?>">
-            <?php echo $item->text; ?>
+            <?php echo $item->name; ?> (<?php echo $item->numarticles; ?>)
```
Comment (laoneo) [inline modules/mod_articles_archive/tmpl/default.php]: I like this, but can you revert this change and make it then against 5.0? Output should stay on 4.4 the same as before.

File: modules/mod_articles_archive/src/Helper/ArticlesArchiveHelper.php [hunk 4]
```diff
@@ -24,73 +28,110 @@
 /**
  * Helper for mod_articles_archive
  *
- * @since  1.5
+ * @since  __DEPLOY_VERSION__
```
Comment (laoneo) [inline modules/mod_articles_archive/src/Helper/ArticlesArchiveHelper.php]: This should be reverted

File: tests/System/integration/site/modules/mod_articles_archive/Default.cy.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+describe('Test the articles archive module', () => {
+  it('can load in frontend and showing a list of months with archived articles', () => {
+    cy.db_createArticle()
+      .then(() => cy.db_createModule({ module: 'mod_articles_archive' }))
+      .then(() => {
+        cy.visit('/');
+
+        cy.contains('li', 'January, 2023');
```
Comment (laoneo) [inline tests/System/integration/site/modules/mod_articles_archive/Default.cy.js]: It would be good if you can actually test if an archived article is also displayed, similar to what we do here https://github.com/joomla/joomla-cms/blob/4.4-dev/tests/System/integration/site/modules/mod_articles_latest/Default.cy.js#L8

File: modules/mod_articles_archive/src/Helper/ArticlesArchiveHelper.php [hunk 6]
```diff
@@ -93,4 +99,27 @@ public static function getList(&$params)
...
     }
+
+    /**
+     * Retrieve list of archived articles
+     *
+     * @param   Registry  &$params module parameters
+     *
+     * @return  \stdClass[]
+     *
+     * @since   1.6
```
Comment (laoneo) [inline modules/mod_articles_archive/src/Helper/ArticlesArchiveHelper.php]:
```suggestion
     * @since   1.5
```

File: modules/mod_articles_archive/services/provider.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_articles_archive
+ *
+ * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (laoneo) [inline modules/mod_articles_archive/services/provider.php]:
```suggestion
 * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
```

File: modules/mod_articles_archive/services/provider.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Registers the service provider with a DI container.
+     *
+     * @param   Container  $container  The DI container.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function register(Container $container)
```
Comment (laoneo) [inline modules/mod_articles_archive/services/provider.php]:
```suggestion
    public function register(Container $container): void
```

File: modules/mod_articles_archive/src/Dispatcher/Dispatcher.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  mod_articles_archive
+ *
+ * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
```
Comment (laoneo) [inline modules/mod_articles_archive/src/Dispatcher/Dispatcher.php]:
```suggestion
 * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
```

File: modules/mod_articles_archive/src/Dispatcher/Dispatcher.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * @return  array
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function getLayoutData(): array
+    {
+        $data = parent::getLayoutData();
+
+        $data['list'] = $this->getHelperFactory()->getHelper('ArticlesArchiveHelper', $data)->getArticlesByMonths($data['params'], $data['app']);
```
Comment (laoneo) [inline modules/mod_articles_archive/src/Dispatcher/Dispatcher.php]: Is the $data attribute really needed in the getHelper call?

**Comment by laoneo** [comment]
Thank you very much, the end result is really nice!

**Comment by carlitorweb** [comment]
@laoneo you are welcome. Really sorry for all the extra-work.

**Comment by laoneo** [comment]
No problem, next time will be much easier. I'm glad that people do step up and help me converting core extensions.

File: modules/mod_articles_archive/src/Dispatcher/Dispatcher.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+namespace Joomla\Module\ArticlesArchive\Site\Dispatcher;
+
+use Joomla\CMS\Dispatcher\AbstractModuleDispatcher;
+use Joomla\CMS\Helper\HelperFactoryAwareInterface;
+use Joomla\CMS\Helper\HelperFactoryAwareTrait;
+
+// phpcs:disable PSR1.Files.SideEffects
+\defined('JPATH_PLATFORM') or die;
```
Comment (sandewt) [inline modules/mod_articles_archive/src/Dispatcher/Dispatcher.php:18]: One more comment though after merging. To be consistent, use here:

`\defined('_JEXEC') or die;`

Comment (laoneo) [reply modules/mod_articles_archive/src/Dispatcher/Dispatcher.php:18]: Thanks for the hint, can you make a new pr for this?

**Comment by sandewt** [comment]
> Thanks for the hint, can you make a new pr for this?

I noticed even more details in the already modified modules. I think it would be useful to change those as well, so that they become more consistent. But first gain some more experience with the modules.
